### PR TITLE
[inductor] [cpp] improve cache blocking with CPU info

### PR DESCRIPTION
**Pull request by chunyuan-w**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #130821
* #131024
* __->__ #129348
* #130690
* #130675

## Description
For single thread case, this PR improves the cache blocking in CPP GEMM template with the CPU info (the L1 and L2 cache size). `Mc_blocks` and `Kc_blocks` are calculated based on the below condition:
     - size_of_B < L1
     - size_of_A < 0.5 * L2

For multi-thread, we need to tune the task decomposition among threads together with cache blocking. We disabled the cache blocking change for now and will submit a follow-up PR for multi-thread optimizations.

## Performance
No regressions. Models with > 3% performance speedup are listed below:

### BF16 single thread (measured on CPU with AMX support)
- static shape

| Model Family | Model Name | Speedup |
|--------------|------------|---------|
torchbench | detectron2_fasterrcnn_r_101_dc5| 4%

- dynamic shape

| Model Family | Model Name | Speedup |
|--------------|------------|---------|
torchbench | detectron2_fasterrcnn_r_101_dc5| 4%

### FP32 single thread (measured on Ice Lake)
- static shape

| Model Family | Model Name | Speedup |
|--------------|------------|---------|
torchbench | basic_gnn_edgecnn| 10%

- dynamic shape

| Model Family | Model Name | Speedup |
|--------------|------------|---------|
torchbench | basic_gnn_edgecnn| 10%

### Next step
The E2E level improvement is limited due to the below reasons:

- For several HF models, we can observe performance improvement at kernel level for the gemm template kernel but since the performance is either still worse than ATen kernel (thus won't be selected during autotune) or improved from worse than ATen to similar to ATen, so we don't see E2E level performance change.

- There're models where the gemm template kernel could get > 10% performance improvement with this PR but since the kernel time is only about 3% of the E2E time, we don't observe significant E2E level improvement.

We will continue to find possible optimizations in the gemm template kernel in follow-up PRs.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/codegen/cpp_micro_gemm.py [hunk 1]
```diff
@@ -542,7 +542,7 @@ class CppMicroGemmAMX(CppMicroGemm):
         {%- set tile_idx_b = tile_offset_b + tile_col %}
         {%- set tile_idx_c = tile_row * num_columns + tile_col %}
         {%- if tile_col == 0 %}
-        _tile_loadd({{tile_idx_a}}, A + {{tile_row * 16}} * lda + k, lda * sizeof({{input_t}}));
+        _tile_stream_loadd({{tile_idx_a}}, A + {{tile_row * 16}} * lda + k, lda * sizeof({{input_t}}));
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp_micro_gemm.py:545]: Can we separate this change to another PR? Or the perf gain from this depends on cache blocking?

Comment (chunyuan-w) [reply torch/_inductor/codegen/cpp_micro_gemm.py]: Sure, moved this change to another PR: https://github.com/pytorch/pytorch/pull/129455

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: torch/_inductor/codegen/cpp_gemm_template.py [hunk 2]
```diff
@@ -220,12 +220,62 @@ def get_blocking(num_threads, factor, m_blocks, n_blocks, k_blocks):
...
+            L2_cache_size = (
+                torch._C._cpu._L2_cache_size()
+            )  # per core cache size in Bytes
+            assert (
+                L2_cache_size > 0
+            ), f"Expect L2_cache_size > 0 but got {L2_cache_size}"
+            B_size_limit = L1_cache_size * L1_limit_factor
+            A_size_limit = L2_cache_size * L2_limit_factor
+
+            num_byte = torch.tensor([], dtype=self.layout.dtype).element_size()
```
[the remark this answers is not in the document]
Comment (leslie-fang-intel) [reply torch/_inductor/codegen/cpp_gemm_template.py:254]: I think int8 should be same here. A separate question is do we need to consider of the VNNI Size here?

[the remark this answers is not in the document]
Comment (jgong5) [reply torch/_inductor/codegen/cpp_gemm_template.py:254]: > I think int8 should be same here.

I thought the dtype of A/B might be different from the dtype in "layout" for int8, no?

[the remark this answers is not in the document]
Comment (leslie-fang-intel) [reply torch/_inductor/codegen/cpp_gemm_template.py:254]: oh, yes, `self.layout` should return `torch.int32` for int8 case which should be wrong.

[the remark this answers is not in the document]
Comment (chunyuan-w) [reply torch/_inductor/codegen/cpp_gemm_template.py]: Thanks for pointing this out. Let me update the code to use `self.input_nodes[0].get_dtype()` for A and `self.input_nodes[1].get_dtype()` for B.

File: torch/_inductor/codegen/cpp_gemm_template.py [hunk 3]
```diff
@@ -220,12 +220,62 @@ def get_blocking(num_threads, factor, m_blocks, n_blocks, k_blocks):
...
+
+            size_cache_B = K0 * Kc_blocks * N0 * num_byte
+
+            import math
```
[the remark this answers is not in the document]
Comment (chunyuan-w) [reply torch/_inductor/codegen/cpp_gemm_template.py]: Changed to a global import

File: torch/_inductor/codegen/cpp_gemm_template.py [hunk 4]
```diff
@@ -220,12 +220,62 @@ def get_blocking(num_threads, factor, m_blocks, n_blocks, k_blocks):
...
+
+            def round_down(f):
+                return math.floor(f)
```
[the remark this answers is not in the document]
Comment (chunyuan-w) [reply torch/_inductor/codegen/cpp_gemm_template.py]: Changed to directly calling `math.floor`

File: torch/_inductor/codegen/cpp_gemm_template.py [hunk 5]
```diff
@@ -220,12 +220,62 @@ def get_blocking(num_threads, factor, m_blocks, n_blocks, k_blocks):
...
+            return Mc_blocks, Kc_blocks
+
         assert (
             not self.is_dynamic_M
         ), "Unable to determine cache blocking for dynamic M."
+        register_blocking = self.register_blocking
         thread_blocking = self.thread_blocking()
-        return GemmBlocking(thread_blocking.block_m, 1, thread_blocking.block_k)
+
+        Mc_blocks, Kc_blocks = get_cache_blocking(register_blocking, thread_blocking)
```
[the remark this answers is not in the document]
Comment (chunyuan-w) [reply torch/_inductor/codegen/cpp_gemm_template.py]: Updated `get_cache_blocking` to return `Nc_blocks` as well.

File: aten/src/ATen/cpu/Utils.cpp [hunk 6]
```diff
@@ -75,4 +76,43 @@ bool init_amx() {
...
+  if (!cache) {
+    return 0;
+  }
+  return cache->size;
+#else
+  return 0;
+#endif
+}
+
+uint32_t L1_cache_size() {
```
[the remark this answers is not in the document]
Comment (chunyuan-w) [reply aten/src/ATen/cpu/Utils.cpp]: Renamed as suggested

**Review by jgong5 — APPROVED** [review]

**Review by jansel — APPROVED** [review]
